### ISSUE TITLE
To many and doubled Android notifications

### ISSUE THREAD

**Issue by Lakoja** [tuskyapp/Tusky#3756]
Since #3626 every notification is shown (sent to the notification manager) as a single Android notification.
Only afterwards a summary notification is created which should remove (group) all single notifications.

However this does not work with Android 11 and lower.
Here all these notifications (single and summary are shown).

![grafik](https://github.com/tuskyapp/Tusky/assets/1618905/8c85b5d9-7744-4332-ba6c-d4e8db3ac8cd)

Two smaller issues:
* If a summary notification is about to be shown it does not make much sense for the user to show (potentially many) single notifications beforehand. They are especially shown only every second (rate-limited). It would be better to just show the final summary notification.
* The summary notifications are split by type (favorite, boost, ...) so often multiple are shown. However with Android 11 this only shows the title(s) "5 new interactions" and "2 new interactions" with my example and some names. Even with Android 12 this split is probably confusing but with Android 11 it makes no sense.

* * * *
- Tusky Version: 22
- Android Version: <= 11

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by nikclayton** [comment]
Do they get re-grouped when you close the notification shade? I've observed a bug in Android where if you have the shade down as notifications are being created, and then create the summary they are not collapsed in to the summary.

But closing the notification shade causes them to be grouped, and then reopening the notification shade correctly shows them as grouped.

You mentioned Android 11 and lower, but https://developer.android.com/develop/ui/views/notifications/group says:

> Note: If your app sends four or more notifications and doesn't specify a group, the system [automatically groups](https://developer.android.com/develop/ui/views/notifications/group#automatic-grouping) them on Android 7.0 and higher.

> If a summary notification is about to be shown it does not make much sense for the user to show (potentially many) single notifications beforehand. They are especially shown only every second (rate-limited). It would be better to just show the final summary notification.

I agree with you. But my understanding of https://developer.android.com/develop/ui/views/notifications/group#group-summary (and the example code that they show there) is that the individual notifications must be created first, and then the summary notification is created.

> The summary notifications are split by type (favorite, boost, ...) so often multiple are shown. However with Android 11 this only shows the title(s) "5 new interactions" and "2 new interactions" with my example and some names. Even with Android 12 this split is probably confusing but with Android 11 it makes no sense.

According to https://developer.android.com/develop/ui/views/notifications/group#group-summary this is on API 23 and below. 

> The summary notifications are split by type (favorite, boost, ...) so often multiple are shown.

You have to do that, because the user might have different settings for the different notification channels, but Tusky can't know what those are set to (Android doesn't make that info available to apps).

There's more detail about this in https://write.as/nikclayton/tusky-v22-0-day-4-android-notifications

**Comment by nikclayton** [comment]
I don't see the duplicate notifications either. This is a screenshot of an Android 11 (API 30) emulator signed in to my account when I've received multiple notifications:

<img src="https://github.com/tuskyapp/Tusky/assets/773100/f4c141d9-4369-4a71-ab3d-543d8dec60b9" width="65%">

What API level was the emulator you took your screenshot running at?

**Comment by Lakoja** [comment]
> You mentioned Android 11 and lower, but https://developer.android.com/develop/ui/views/notifications/group says:

> I agree with you. But my understanding of https://developer.android.com/develop/ui/views/notifications/group#group-summary (and the example code that they show there) is that the individual notifications must be created first, and then the summary notification is created.

The (second) code example there sets a group on every single notification.
The summary notification then needs the same group.
Also in our case I always would set `setGroupAlertBehavior(NotificationCompat.GROUP_ALERT_SUMMARY)`.

With these two changes it works as expected for me. The sleep can then be removed.
(Would need some refactoring though as all "notifys" must be concerted.)

The only thing remaining then (next step?) would be that multiple groups would get multiple notification sounds. If possible I would also remove the sound of all but the last (or first).

**Comment by nikclayton** [comment]
> The (second) code example there sets a group on every single notification. The summary notification then needs the same group. 

That's what the existing code does, from what I can see.

`NotificationHelper.newAndroidNotification()` calls `.setGroup` on each notification it creates, and sets the group to the channel ID

`NotificationHelper.updateSummaryNotification()` also calls `.setGroup` on each summary notification that it creates, and also sets the group to the channel ID.

> Also in our case I always would set `setGroupAlertBehavior(NotificationCompat.GROUP_ALERT_SUMMARY)`.

This is also already done, on every notification except the first one in the group, so only the first one in the group will alert.

> With these two changes it works as expected for me.

Can you send the specific changes you've made as a patch/diff? I need to see how what you're talking about differs from what the code currently does, because to me it looks like they're the same thing.

> The sleep can then be removed. (Would need some refactoring though as all "notifys" must be concerted.)

I don't think it can, the comment immediately above the sleep explains why:

```
            // Android will rate limit / drop notifications if they're posted too
            // quickly. There is no indication to the user that this happened.
            // See https://github.com/tuskyapp/Tusky/pull/3626#discussion_r1192963664
            try { Thread.sleep(1000); } catch (InterruptedException ignored) { }
```

Why do you think this is incorrect?

> The only thing remaining then (next step?) would be that multiple groups would get multiple notification sounds. If possible I would also remove the sound of all but the last (or first).

Whether or a not a group makes a sound is ultimately governed by how the user has configured the channel.

The app can't know what those settings are. So it has no way of knowing whether it's just sent a notification to a channel that will make a sound or not.

Suppose I get 3 notifications, one follow, one mention, one boost, in that order.

I have configured the channels so that follow and boost are silent, mention will make a sound.

Tusky can't know that. So if it decides that only the first (follow), or last (boost) notification should make a sound **I will not be alerted to the mention**, and this is the only one I definitely **want** to be alerted about.

**Comment by Lakoja** [comment]
> > The (second) code example there sets a group on every single notification. The summary notification then needs the same group.
> 
> That's what the existing code does, from what I can see.
> 
> > Also in our case I always would set `setGroupAlertBehavior(NotificationCompat.GROUP_ALERT_SUMMARY)`.
> 
> This is also already done, on every notification except the first one in the group, so only the first one in the group will alert.

You seem to be right. It behaves again differently today.
(I also seem to be having difficulties understanding the code, even on day three of work there..)

One smaller detail: the `isFirstOfBatch` shouldn't that be a "first of group"?
Notifications in the processed list are not grouped for the moment so the very first one would sound and after that any group notification which is not the group of the first one. (at least that is how it behaves for me)

> > The sleep can then be removed. (Would need some refactoring though as all "notifys" must be concerted.)
> 
> I don't think it can, the comment immediately above the sleep explains why:

I would still be surprised if that is what one needs to do in an Android app to show a grouped notification:
First schedule 10 Android notifications across 10 seconds (is that delay long enough?) and only then group them.
(The code example above schedules its three notifications in one call (`apply`).)

Are there other apps out there that do it like this? Or is Tusky unique that it actually has multiple notifications in one moment (pull notifications with 15 minutes delay) whereas every other app has instant notifications so when it sounds twice its actually two distinct interactions?

**Comment by Lakoja** [comment]
> As an example: I posted something yesterday which resulted in quite a few reposts and favourites from various users. The majority of these notifications is 12+ hours ago, yet I still receive some on the Tusky app right now...
> 
> I'm using ntfy for faster push notifications and Tusky did create endpoints for these push notificiations, yet they still come with often a really big delay...
> 
> Edit: Looking at it, I think the now 13+ hours are related to the post itself? Would make sense...

Yes, the timestamps in the notification relate to the original post. (That confuses me from time to time.)
If **push** notifications are used and they are delayed I guess this might also be due to the push distributor (here ntfy).
I had a quite popular one last week and notifications did stop after 400 or so and resumed on the next day.

The issue here is only with **pull** notifications.

**Comment by nikclayton** [comment]
> I would still be surprised if that is what one needs to do in an Android app to show a grouped notification:
> First schedule 10 Android notifications across 10 seconds (is that delay long enough?) and only then group them.
> (The code example above schedules its three notifications in one call (apply).)
> 
> Are there other apps out there that do it like this? Or is Tusky unique that it actually has multiple notifications in one moment 
> (pull notifications with 15 minutes delay) whereas every other app has instant notifications so when it sounds twice its actually 
> two distinct interactions?

That's probably one of the reasons Tusky can create more notifications at once than other apps.

I researched this before I wrote the code, and there's a link to it in the code comments.

Here it is again -- https://github.com/tuskyapp/Tusky/pull/3626.

You should be able to reproduce this experiment and confirm for yourself that if you don't rate limit notifications you will see log messages indicating that Android is muting and/or dropping some notifications because they are being posted too quickly.

**Comment by nikclayton** [comment]
Also, as noted in a comment in the code, there is an API that supports batch sending multiple notifications at once (https://developer.android.com/reference/androidx/core/app/NotificationManagerCompat#notify(java.util.List%3Candroidx.core.app.NotificationManagerCompat.NotificationWithIdAndTag%3E)), and we can switch to using that when https://developer.android.com/jetpack/androidx/releases/core 1.11 is released (currently on 1.11.0-beta02)

**Comment by Lakoja** [comment]
After further testing I can add the following:

* The **grouping works**: notifications for one account and one type get the same group id and afterwards a summary notification with that group id
* From a rate limit perspective with the **delay removed**: I do only get a "count" log (not a rate limit): "Package has already posted or enqueued 25 notifications.  Not showing more.  package=com.keylesspalace.tusky"
* The "**firstOfBatch**" logic should be fixed to a "firstOfGroup"; for that a grouping of the single notifications by "group id" should be done before the `notify()` loop
* It is then still odd to see a (potentially) large number of notifications rushing in visually but that can later be fixed by the "**batch sending**"
* I might add another issue with the number of notifications channels. For example for 2 accounts it seems excessive to then have 20 single **notification settings** in the Android settings? (This would potentially solve my problem with "still too many notifications because different groups".)

**Comment by nikclayton** [comment]
> For example for 2 accounts it seems excessive to then have 20 single notification settings in the Android settings?

How many settings do you think there should be, and which of the existing settings do you propose removing?

**Comment by Lakoja** [comment]
> > For example for 2 accounts it seems excessive to then have 20 single notification settings in the Android settings?
> 
> How many settings do you think there should be, and which of the existing settings do you propose removing?

(also see the small screenshot in the description above)

Separating between Android notifications (icon in notification bar) and Tusky notifications:

Other applications nearly always have only one (channel). This means they also only have one notification icon showing in the notification bar.
Which seems correct to me: One icon means "something new in this app".
Several icons like Tusky is doing now doesn't tell you more but clutters the space.
(Also note that for example the messages app changes the single notification icon to a "there are several things" icon if there are several new messages.)

Having said that it would probably still be necessary to configure which notification types in Tusky will be shown as Android notification. This would have to be Tusky-internal. But this could for example reflect the filter setting of the notification timeline (would also be simpler, less confusing, less settings for the user, also remove a confusing jump to external settings from the app settings).

**Comment by annando** [comment]
Concerning this "too many notifications": It is quite irritating when your whole notification screen looks like this:
![Screenshot_20230814-084107](https://github.com/tuskyapp/Tusky/assets/844208/225a51d3-bfff-4c8c-abc6-181815297e3d)
Also this behaviour seems to be responsible for multiple consequent notifications by vibration - which is quite irritating.

My device is running Android 13.

**Comment by nikclayton** [comment]
@annando That should be one group per notification type (mention, follow, boost, etc).

You can disable those on a per-type basis at Account preferences > Notifications. E.g., you can use this to show only the types of notifications that you care about.

That is also where you can control whether the notifications vibrate, make a sound, etc.

**Comment by annando** [comment]
Isn't there any option to define that all the different notification types should be grouped into a single notification? Also it seems as if new notifications are just added up without removing/editing of the existing notifications? I don't know if this technically possible, since I'm not into Android development. But having 10 notifications by a single app is quite intense. I really would prefer having just a single notification for everything.

**Comment by Lakoja** [comment]
> But having 10 notifications by a single app is quite intense. I really would prefer having just a single notification for everything.

That was the case before Tusky v22 - grouping by Tusky.
However all of them should now be grouped by Android into a single one (with multiple lines).
However I did start the issue because I was seeing 30 separate ones; that specific problem I couldn't reproduce afterwards anymore.

Do you have this problem every time?
What Android version is this?

**Comment by annando** [comment]
@Lakoja I'm using Tusky 23.0 with Android 13 - and it happens all the time.

**Comment by Lakoja** [comment]
> @Lakoja I'm using Tusky 23.0 with Android 13 - and it happens all the time. See my screenshot here: [#3756 (comment)](https://github.com/tuskyapp/Tusky/issues/3756#issuecomment-1677462144)

Thanks. Which device is that?

**Comment by annando** [comment]
It's a Fairphone 3+

**Comment by nikclayton** [comment]
@annando 

> Isn't there any option to define that all the different notification types should be grouped into a single notification?

No. All notifications *of the same type* can be grouped in to a single notification (which is what the Tusky code should do).

Notifications of different types can't be grouped together, because then it wouldn't be possible to set different notification styles (make a sound, vibrate, bypass do-not-disturb) on different types of notification, or swipe away all notifications of one type without also swiping away all notifications.

> Also it seems as if new notifications are just added up without removing/editing of the existing notifications

Yes. Tusky tries not to remove a notification unless the app can be sure that you've seen it.  A notification is removed if:

1. You dismiss it manually (e.g., swipe, so you must have seen it).
2. You open the "Notifications" tab in the app -- all active Android notifications should be removed because Tusky knows that you are looking at the notifications now
3. You have more than ~ 40 active notifications. In which case adding a new notification will remove the oldest one, because ~ 40 is the limit on active notifications per-app imposed by Android.

If you don't want to see (or hear) particular types of notifications then "Account preferences > Notifications" will let you configure them on a type-by-type basis.

**Comment by annando** [comment]
I'm not the type of person who looks at their phone every few seconds/minutes, especially when I'm at work. So it can happen that I pick it up after maybe 4 hours. Sadly, this current behaviour renders the notification functionality useless for me. I want to be notified, but for me it would be enough if there even was a simple "look into your app" notification that doesn't goes off every time something new arrives. But I guess that this is more a niche requirement and I understand why there is no way enabling such a functionality. Sadly I never coded an Android app, so I won't be able providing some PR for this.

**Comment by Lakoja** [comment]
> I want to be notified, but for me it would be enough if there even was a simple "look into your app" notification that doesn't goes off every time something new arrives. But I guess that this is more a niche requirement 

Sounds sensible at least to me.
I guess even on Mastodon most users are no hardcore-I-need-to-configure-everything types.
And even if there is truely much and interesting new stuff, anyone would want to check that in the app (with more space and filtering and stuff) and not only as fine-grained notifications.
